lib/src/main/java/com/couchbase/lite/NativeLibrary.java

```diff
@@ -23,54 +23,82 @@
 
+    /**
+     * Extracts and loads native libraries.
+     */
     static void load() {
         if (LOADED.getAndSet(true)) { return; }
-
-        for (String lib : LIBRARIES) { load(lib); }
-    }
-
-    private static void load(String libName) {
         try {
-            final File libFile = extractLibrary(libName);
-            System.load(libFile.getAbsolutePath());
+            final String[] libPaths = Arrays.stream(LIBRARIES).map(lib -> getResourcePath(lib)).toArray(String[]::new);
+            final String targetDir = getTargetDirectory(libPaths);
+            for (String path : libPaths) {
+                System.load(extract(path, targetDir).getAbsolutePath());
+            }
         }
         catch (Throwable th) {
             final String platform = System.getProperty("os.name") + "/" + System.getProperty("os.arch");
-            throw new IllegalStateException("Cannot load native library : " + libName + " for " + platform, th);
+            throw new IllegalStateException("Cannot load native library for " + platform, th);
         }
     }
 
+    /**
+     * Returns a target directory for extracting the native libraries into. The structure of the
+     * directory will be <System Temp Directory>/com.couchbase.lite.java/native/<MD5-Hash>.
+     * The MD5-Hash is the combined MD5 hash of the hashes of all native libraries.
+     */
+    @NonNull
+    private static String getTargetDirectory(@NonNull String[] libPaths) throws Exception {
+        final MessageDigest md = MessageDigest.getInstance("MD5");
+        for (String path : libPaths) {
+            final URI md5File = NativeLibrary.class.getResource(path + ".MD5").toURI();
+            md.update(Files.readAllBytes(Paths.get(md5File)));
+        }
+        final String md5 =  String.format("%032x", new BigInteger(1, md.digest()));
+        return new File(CouchbaseLite.getTmpDirectory(TARGET_BASE_DIR), md5).getAbsolutePath();
+    }
+
+    /**
+     * Extracts the given path to the native library in the resource directory into the target directory.
+     * If the native library already exists in the target library, the existing native library will be used.
+     */
     @NonNull @SuppressFBWarnings("DE_MIGHT_IGNORE")
-    private static File extractLibrary(@NonNull String libName) throws IOException, InterruptedException {
-        final String libResPath = getLibraryResourcePath(libName);
-        final String tmpDir = CouchbaseLite.getTmpDirectory(TMP_DIR_NAME);
-        final String targetDir = new File(tmpDir).getAbsolutePath();
-        final File targetFile = new File(targetDir, System.mapLibraryName(libName));
-
-        if (targetFile.exists() && !targetFile.delete()) {
-            throw new IllegalStateException("Failed to delete the existing native library at " +
-                        targetFile.getAbsolutePath());
+    private static File extract(@NonNull String libResPath, @NonNull String targetDir) throws Exception {
```

Comment:
I think this just throws IOException, no?

Reply:
Yes, they are IOException and InterruptedException. Will add those to the method signatures.